File: solutions/0680. Valid Palindrome II/0680.cs
// Time complexity: O(n)
// Space complexity: O(1)
public class Solution {
    public bool ValidPalindrome(string s) {
        for (int l = 0, r = s.Length - 1; l < r; ++l, --r) {
            if (s[l] != s[r]) {
                return ValidPalindrome(s, l + 1, r) || ValidPalindrome(s, l, r - 1);
            }
        } 
        return true;
    }

    private bool ValidPalindrome(string s, int l, int r) {
        while (l < r) {
            if (s[l++] != s[r--]) {
                return false;
            }
        }
        return true;
    }
}

File: solutions/0680. Valid Palindrome II/0680.js
// Time complexity: O(n)
// Space complexity: O(1)
var validPalindrome = function(s) {
    let l = 0;
    let r = s.length - 1;
  
    while (l < r) {
        if (s[l] !== s[r]) {
            return isPalindrome(s, l + 1, r) || isPalindrome(s, l, r - 1);
        }
        l++;
        r--;
    }
  
    return true;

    function isPalindrome(s, left, right) {
        while (left < right) {
            if (s[left] !== s[right]) {
                return false;
            }
            left++;
            right--;
        }
        return true;
    }
};